core/src/apps/cardano/address.py

```diff
@@ -1,137 +1,249 @@
+
+    address_id = _get_address_id(address_type)
+    header = address_id << 4 | network_id
 
-def _address_hash(data) -> bytes:
-    data = cbor.encode(data)
-    data = hashlib.sha3_256(data).digest()
-    res = hashlib.blake2b(data=data, outlen=28).digest()
-    return res
+    return bytes([header])
 
 
-def _get_address_root(node, payload):
-    extpubkey = remove_ed25519_prefix(node.public_key()) + node.chain_code()
-    if payload:
-        payload = {1: cbor.encode(payload)}
+def _get_address_id(address_type: int) -> int:
+    if address_type == CardanoAddressType.BYRON_ADDRESS:
+        address_id = AddressId.BYRON_ADDRESS
+    elif address_type == CardanoAddressType.BASE_ADDRESS:
+        address_id = AddressId.BASE_ADDRESS
+    elif address_type == CardanoAddressType.POINTER_ADDRESS:
+        address_id = AddressId.POINTER_ADDRESS
+    elif address_type == CardanoAddressType.ENTERPRISE_ADDRESS:
+        address_id = AddressId.ENTERPRISE_ADDRESS
+    elif address_type == CardanoAddressType.REWARD_ADDRESS:
+        address_id = AddressId.REWARD_ADDRESS
     else:
-        payload = {}
-    return _address_hash([0, [0, extpubkey], payload])
+        raise ValueError("Invalid address type '%s'!" % address_type)
+
+    return address_id
+
+
+def _derive_base_address(
+    keychain: seed.Keychain,
+    address_header: bytes,
+    path: List[int],
+    staking_key_path: List[int],
+    staking_key_hash: bytes,
+) -> str:
+    spending_key_hash = get_public_key_hash(keychain, path)
+
+    _validate_base_address_staking_info(staking_key_path, staking_key_hash)
+
+    if staking_key_hash is None:
+        staking_key_hash = get_public_key_hash(keychain, staking_key_path)
+
+    return address_header + spending_key_hash + staking_key_hash
+
+
+def _validate_base_address_staking_info(
+    staking_key_path: List[int], staking_key_hash: bytes,
+):
+    if staking_key_hash is None:
+        if staking_key_path is None:
+            raise wire.DataError(
+                "Base address needs a staking key path or a staking key hash!"
+            )
+
+        if not _is_staking_path(staking_key_path):
+            raise wire.DataError("Invalid staking key path!")
+
+
+def _is_staking_path(path: List[int]) -> bool:
+    return validate_full_path(path) and path[3] == 2
+
+
+def _derive_pointer_address(
+    keychain: seed.Keychain,
+    address_header: bytes,
+    path: List[int],
+    pointer: CardanoBlockchainPointerType,
+) -> str:
+    spending_key_hash = get_public_key_hash(keychain, path)
+
+    encoded_pointer = _encode_certificate_pointer(pointer)
+
+    return address_header + spending_key_hash + encoded_pointer
+
+
+def _encode_certificate_pointer(pointer: CardanoBlockchainPointerType) -> bytes:
+    if (
+        pointer is None
+        or pointer.block_index is None
+        or pointer.tx_index is None
+        or pointer.certificate_index is None
+    ):
+        raise wire.DataError("Invalid pointer!")
```

Comment:
Resolved offline - I don't like the idea of having duplicated logic because of early validation. So we'll keep it like this and we'll see if Trezor guys share the same concerns.